File: packages/borg-sdk-py/src/borg/protocol.py
"""
The wire messages, as this side of the connection sees them. ``crates/borg-protocol`` is the
contract; this is a transcription of it, and the scenarios are what keep the two honest.

**Every message is a single-key object**, including the payload-free ones, so dispatching is one
key lookup with no special cases (§17.4).

Engine → worker::

    {"invoke": {"producer": "12342029420047889112", "input": "Company:o-04068"}}
    {"value": "acme.ai"}   # or {"value": null} — the cell has never been written
    {"ok": {}}
    {"shutdown": {}}

Worker → engine::

    {"get": "Company:o-04068.website"}
    {"get_input": "Company:o-04068.founded"}     # migrations only (§9.3)
    {"set": {"cell": "Company:o-04068.isInvestible", "value": "true"}}
    {"done": {}}
    {"error": {"message": "…"}}

The messages are plain dicts rather than classes. There is nothing to model: each one is a single
key whose value is a string, ``None``, or a two-key object, and a dataclass layer over that would be
a second place for the shape to drift from ``borg-protocol``.
"""

from __future__ import annotations

__all__ = ["DESCRIBE_ARG", "TRANSPORT", "VERSION", "producer_id"]

#: The protocol version the engine announces in its hello.
VERSION = 1

#: The transport every repo written with this SDK asks for. See :mod:`borg.connection`.
TRANSPORT = "socket"

#: The one argument the engine passes an executable before it is ever a worker.
DESCRIBE_ARG = "describe"


def producer_id(name: str) -> str:
    """
    The id the engine derives from a producer's name (§9.2), computed here so a repo serving several
    pipelines can dispatch an ``invoke`` to the right one.

    FNV-1a over 64 bits. It is returned **as a string**, and the engine sends it as one, because a
    ``ProducerId`` is a hash past 2⁵³ and JSON has no integers: a client that read it as a JSON
    number would resolve a producer that does not exist. Python's ``int`` could hold it perfectly
    well — the string is not for Python's benefit, it is the contract, and a dispatch table keyed on
    a different type in each language is a bug waiting for the first repo with two pipelines.
    """
    mask = (1 << 64) - 1
    hash_ = 0xCBF29CE484222325
    for byte in name.encode("utf-8"):
        hash_ = (hash_ ^ byte) & mask
        hash_ = (hash_ * 0x100000001B3) & mask
    # Kept away from the small ids a human might type into a def file, exactly as the engine does.
    return str(hash_ | (1 << 32))

File: packages/borg-sdk/src/protocol.ts
/**
 * The wire messages, as this side of the connection sees them. `crates/borg-protocol` is the
 * contract; this is a transcription of it, and the scenarios are what keep the two honest.
 *
 * **Every message is a single-key object**, including the payload-free ones, so dispatching is one
 * property lookup with no special cases (§17.4).
 */

export interface ServerHello {
  version: number;
  codecs: string[];
}

export type ToWorker =
  | { invoke: { producer: string; input: string } }
  | { value: string | null }
  | { ok: Record<string, never> }
  | { shutdown: Record<string, never> };

export type FromWorker =
  | { get: string }
  | { get_input: string }
  | { set: { cell: string; value: string } }
  | { done: Record<string, never> }
  | { error: { message: string } };

/** What a repo reports when asked to `describe` itself (§17.4). */
export interface Description {
  structs: StructSpec[];
  producers: ProducerSpec[];
  migrations?: MigrationSpec[];
  transport: "stdio" | "socket";
  repo?: number;
}

export interface StructSpec {
  name: string;
  fields: FieldSpec[];
}

export interface FieldSpec {
  name: string;
  type: string;
  derived_by?: string;
  up?: string;
  down?: string;
}

export interface ProducerSpec {
  name: string;
  source: string;
}

export interface MigrationSpec {
  name: string;
}

/**
 * The id the engine derives from a producer's name (§9.2), computed here so a repo serving several
 * pipelines can dispatch an `invoke` to the right one.
 *
 * FNV-1a over 64 bits, in `bigint` because that is the only JS type that can hold the answer. The
 * engine sends the id **as a string** for the same reason: read as a JSON number it would round to
 * 53 bits and name a producer that does not exist.
 */
export function producerId(name: string): string {
  const MASK = (1n << 64n) - 1n;
  let hash = 0xcbf29ce484222325n;
  for (const byte of new TextEncoder().encode(name)) {
    hash = (hash ^ BigInt(byte)) & MASK;
    hash = (hash * 0x100000001b3n) & MASK;
  }
  // Kept away from the small ids a human might type into a def file, exactly as the engine does.
  return String(hash | (1n << 32n));
}
